File: leetcode/sliding-window/0076_minimum_window_substring.py
# LC 76. Minimum Window Substring | Hard
from collections import Counter

class Solution:
    def minWindow(self, s, t):
        if not t or not s: return ""
        need = Counter(t)
        have, required = 0, len(need)
        res, res_len = [-1, -1], float('inf')
        l, window = 0, {}
        for r, ch in enumerate(s):
            window[ch] = window.get(ch, 0) + 1
            if ch in need and window[ch] == need[ch]: have += 1
            while have == required:
                if (r - l + 1) < res_len:
                    res, res_len = [l, r], r - l + 1
                window[s[l]] -= 1
                if s[l] in need and window[s[l]] < need[s[l]]: have -= 1
                l += 1
        return s[res[0]:res[1]+1] if res_len != float('inf') else ""

File: leetcode/sliding-window/0076_minimum_window_substring.cpp
// LC 76. Minimum Window Substring | Hard
#include <string>
#include <unordered_map>
#include <climits>
using namespace std;

class Solution {
public:
    string minWindow(string s, string t) {
        unordered_map<char,int> need, window;
        for (char c : t) need[c]++;
        int have = 0, required = need.size(), l = 0, minLen = INT_MAX, start = 0;
        for (int r = 0; r < s.size(); r++) {
            window[s[r]]++;
            if (need.count(s[r]) && window[s[r]] == need[s[r]]) have++;
            while (have == required) {
                if (r - l + 1 < minLen) { minLen = r - l + 1; start = l; }
                window[s[l]]--;
                if (need.count(s[l]) && window[s[l]] < need[s[l]]) have--;
                l++;
            }
        }
        return minLen == INT_MAX ? "" : s.substr(start, minLen);
    }
};
